### PR TITLE
Tftp backport/v1

### PR DESCRIPTION
**Pull request by shivan1b**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) tickets:
- https://redmine.openinfosecfoundation.org/issues/3044
- https://redmine.openinfosecfoundation.org/issues/3008

### CODE REVIEW

**Comment by victorjulien** [comment]
https://ci.appveyor.com/project/OISF/suricata/builds/25658705

Rust binding included for non-Rust build. Probably a side effect of git master no longer requiring ```#ifdef RUST``` guards, but master-4.1.x still does.

File: rust/src/dhcp/dhcp.rs [hunk 1]
```diff
@@ -239,21 +240,21 @@ pub extern "C" fn rs_dhcp_probing_parser(_flow: *const Flow,
 
 #[no_mangle]
 pub extern "C" fn rs_dhcp_tx_get_alstate_progress(_tx: *mut libc::c_void,
-                                                  _direction: libc::uint8_t) -> libc::c_int {
+                                                  _direction: u8) -> libc::c_int {
```
Comment (jlucovsky) [inline rust/src/dhcp/dhcp.rs]: nit: The `{` should be on the same line or the next line consistently (IMO). This change is inconsistent with that on line 225.

File: rust/src/tftp/tftp.rs [hunk 2]
```diff
@@ -31,11 +31,28 @@ pub struct TFTPTransaction {
...
     pub mode : String,
-    pub logged : LoggerFlags
+    pub logged : LoggerFlags,
+    id: u64,
 }
 
 pub struct TFTPState {
-    pub transactions : Vec<TFTPTransaction>
+    pub transactions : Vec<TFTPTransaction>,
+    /// tx counter for assigning incrementing id's to tx's
```
Comment (jlucovsky) [inline rust/src/tftp/tftp.rs]: nit: use 2 `/` characters

File: src/app-layer-ikev2.c [hunk 3]
```diff
@@ -34,33 +34,9 @@
 
 #include "app-layer-ikev2.h"
 
-#ifdef HAVE_RUST
```
Comment (jlucovsky) [inline src/app-layer-ikev2.c]: Is Rust mandatory in 4.1.4?

File: src/app-layer-tftp.c [hunk 4]
```diff
@@ -36,14 +36,6 @@
 
 #include "app-layer-tftp.h"
 
-#ifndef HAVE_RUST
```
Comment (jlucovsky) [inline src/app-layer-tftp.c]: Rust mandatory?

**Comment by shivan1b** [comment]
@jlucovsky there are no compiler warnings about using `libc::c_void` but yes using `std::os::raw::c_void` seems like a better choice for our usecases. Since this is just a backport of Victor's commits from master, I can make the change of using `std::os::raw::c_void` as a separate cleanup patch if @victorjulien thinks its OK too.

**Comment by shivan1b** [comment]
Please check the new PR with changes requested here: #3986

File: rust/src/tftp/tftp.rs (hunk 2 above)
Comment (victorjulien) [reply rust/src/tftp/tftp.rs]: In Rust ```///``` has a special meaning for documenting, so it may be correct here.

**Pull request closed without merging**